### DISCUSSION TITLE
Building for x64_86 (many_linux)

### DISCUSSION THREAD

**Discussion by why-not-try-calmer** [PyO3/pyo3#3139] [Questions] [answered]
Hello,

I am trying to build wheels for `x64_86` under the "many_linux" naming convention, which is required to upload the wheels to PyPI. I am building very simply, like so:

```bash
python3 -m build
```
The output:
```
my_project-0.1.0-cp310-cp310-linux_x86_64.whl
my_project-0.1.0.tar.gz
```
As mentioned this wheels file _cannot_ be uploaded as is to PyPI; it must be "repaired" to match the [manylinux](https://github.com/pypa/manylinux), which I do with

```bash
docker run -it -v (pwd)/dist:/dist quay.io/pypa/manylinux_2_28_x86_64 bash
auditwheel repair /dist/my_project-0.1.0-cp310-cp310-linux_x86_64.whl /dist/repaired
```

And the issue is that I am getting: `auditwheel: error: cannot repair "/dist/my_project-0.1.0-cp310-cp310-linux_x86_64.whl" to "manylinux_2_28_x86_64" ABI because of the presence of too-recent versioned symbols. You'll need to compile the wheel on an older toolchain.` And indeed, `auditwheel show ...` suggests:

```
The wheel references external versioned symbols in these
system-provided shared libraries: libgcc_s.so.1 with versions
{'GCC_3.3', 'GCC_4.2.0', 'GCC_3.0'}, libc.so.6 with versions
{'GLIBC_2.34', 'GLIBC_2.18', 'GLIBC_2.17', 'GLIBC_2.14', 'GLIBC_2.33',
'GLIBC_2.2.5', 'GLIBC_2.28', 'GLIBC_2.3', 'GLIBC_2.3.4'}

This constrains the platform tag to "manylinux_2_34_x86_64". In order
to achieve a more compatible tag, you would need to recompile a new
wheel from source on a system with earlier versions of these
libraries, such as a recent manylinux image.
```
Thing is, it looks a bit convoluted to build a custom image on top of the manylinux image adding the Rust toolchain just so that I can build from that image the wheels with a lower glibc version. 

Is there a more straightforward approach?

By the way thanks for this amazing library, which I feel I will be using more and more!

**Comment by adamreichold** [comment]
Which build system backend are you using? `setuptools-rust` or `maturin`? Maturin generally has very good support distributable wheels, c.f. https://www.maturin.rs/distribution.html

**Reply by why-not-try-calmer** [reply]
Hello, thanks for your quick reply, I am using `setuptools-rust` as it seemed a bit less-3rd-party dependent than the other.

**Reply by adamreichold** [reply]
> Thing is, it looks a bit convoluted to build a custom image on top of the manylinux image adding the Rust toolchain just so that I can build from that image the wheels with a lower glibc version.

I think using `setuptools-rust`, that would be required but I am not completely sure. The turnkey solution for this is Maturin as far as I know.

**Reply by adamreichold** [reply, marked as answer]
(You could probably use one of the existing Docker images provided by/for Maturin with the setuptools-rust approach, but why not go all the way then.)

**Reply by why-not-try-calmer** [reply]
True. Okay I'll update this convo if find a simple way. Perhaps that'll be worth adding to the docs.
